broker/client.go

```diff
@@ -0,0 +1,647 @@
+	if err != nil {
+		err = errors.Wrap(err, "failure when constructing the broker request token")
+		return
+	}
+	req.Header.Set("Authorization", "Bearer "+token)
+
+	tr := config.GetTransport()
+	client := &http.Client{Transport: tr}
+
+	resp, err := client.Do(req)
+	if err != nil {
+		err = errors.Wrap(err, "Failure when invoking the broker URL")
+		return
+	}
+	defer resp.Body.Close()
+	responseBytes, err := io.ReadAll(resp.Body)
+	if err != nil {
+		err = errors.Wrap(err, "Failure when reading response from broker response")
+	}
+	if resp.StatusCode >= 400 {
+		errResp := brokerErrResp{}
+		log.Errorf("Failure (status code %d) when invoking the broker: %s", resp.StatusCode, string(responseBytes))
+		if err = json.Unmarshal(responseBytes, &errResp); err != nil {
+			err = errors.Errorf("Failure when invoking the broker (status code %d); unable to parse error message", resp.StatusCode)
+		} else {
+			err = errors.Errorf("Failure when invoking the broker (status code %d): %s", resp.StatusCode, errResp.Msg)
+		}
+		return
+	}
+
+	// Connection request sent; create a new host certificate in preparation for a response.
+	serialNumberLimit := new(big.Int).Lsh(big.NewInt(1), 128)
+	serialNumber, err := rand.Int(rand.Reader, serialNumberLimit)
+	if err != nil {
+		return
+	}
+	notBefore := time.Now()
+	template := x509.Certificate{
+		SerialNumber: serialNumber,
+		Subject: pkix.Name{
+			Organization: []string{"Pelican"},
+			CommonName:   originName,
+		},
+		NotBefore:             notBefore,
+		NotAfter:              notBefore.Add(10 * time.Minute),
+		KeyUsage:              x509.KeyUsageDigitalSignature,
+		ExtKeyUsage:           []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth, x509.ExtKeyUsageClientAuth},
+		BasicConstraintsValid: true,
+	}
+	template.DNSNames = []string{originName}
+	derBytes, err := x509.CreateCertificate(rand.Reader, &template, caCert, &privKey.PublicKey, caPrivateKey)
+	if err != nil {
+		return
+	}
+	callbackResp := reversalCallbackResponse{
+		Certificate: base64.StdEncoding.EncodeToString(derBytes),
+	}
+	callbackBytes, err := json.Marshal(&callbackResp)
+	if err != nil {
+		return
+	}
+
+	// Wait for the origin to callback to the cache's return endpoint; that HTTP handler
+	// will write to the channel we originally posted.
+	tck := time.NewTicker(20 * time.Second)
+	log.Debugf("Cache waiting for up to 20 seconds for the origin %s to callback", originName)
+	select {
+	case <-ctx.Done():
+		log.Debug("Context has been cancelled while waiting for callback")
+		err = ctx.Err()
+		return
+	case <-tck.C:
+		log.Debug("Request has timed out when waiting for callback")
+		err = errors.Errorf("Timeout when waiting for callback from origin")
+		return
+	case writer := <-responseChannel:
+		hj, ok := writer.(http.Hijacker)
+		if !ok {
+			log.Debug("Not able to hijack underlying TCP connection from server")
+			resp := brokerErrResp{
+				Msg:    "Unable to reverse TCP connection; HTTP/2 in use",
```

Comment:
Does this mean we can't use `HTTP/2` for reverse connection? I thought we are using `HTTP/2` by default. No?

Reply:
Correct: this doesn't work when HTTP/2 is in use.

I think I perhaps got lucky that the code didn't trip over the problem.  I'll explicitly disable HTTP/2 and make sure existing TCP connections aren't reused.